source/SAMRAI/tbox/MessageStream.h

```diff
@@ -153,6 +153,56 @@ class MessageStream
       return d_buffer_index >= d_buffer_size;
    }
 
+   /*!
+    * @brief Returns a pointer into the message stream valid for
+    * num_bytes.
+    *
+    * @param[in] num_bytes  Number of bytes requested for window.
+    *
+    * @pre readMode()
+    */
+   template<typename DATA_TYPE>
+   const DATA_TYPE *
+   getReadBuffer(
+      size_t num_entries)
+   {
+      TBOX_ASSERT(readMode());
+      const size_t num_bytes = getSizeof<DATA_TYPE>(num_entries);
+      TBOX_ASSERT(canCopyOut(num_bytes));
+      const DATA_TYPE *buffer =
+         reinterpret_cast<const DATA_TYPE *>(&d_read_buffer[getCurrentSize()]);
+      d_buffer_index += num_bytes;
+      return buffer;
+   }
+
+   /*!
+    * @brief Returns a pointer into the message stream valid for
+    * num_bytes.
+    *
+    * @param[in] num_bytes  Number of bytes requested for window.
+    *
+    * @pre writeMode()
+    */
+   template<typename DATA_TYPE>
+   DATA_TYPE *
+   getWriteBuffer(
+      size_t num_entries)
+   {
+      TBOX_ASSERT(writeMode());
+      const size_t num_bytes = getSizeof<DATA_TYPE>(num_entries);
+      if (!growAsNeeded()) {
```

Comment:
This doesn't quite work correctly when d_write_buffer doesn't get resized.  We end up pointing at storage that has been reserved for the vector, but isn't truly part of the vector.  I'll suggest a fix.

Reply:
```
-      if (!growAsNeeded()) {
-         TBOX_ASSERT(canCopyIn(num_bytes));
-      }
-      else if (num_bytes > 0) {
+      if (num_bytes > 0) {
```

Reply:
Fixed!